src/wallet/wallet.cpp

```diff
@@ -2384,6 +2384,23 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
     return true;
 }
 
+void CWallet::CompactDatabase() {
+    WalletDatabase& dbh = *database;
+
+    unsigned int nUpdateCounter = dbh.nUpdateCounter;
+
+    if (dbh.nLastSeen != nUpdateCounter) {
+        dbh.nLastSeen = nUpdateCounter;
+        dbh.nLastWalletUpdate = GetTime();
+    }
+
+    if (dbh.nLastFlushed != nUpdateCounter && GetTime() - dbh.nLastWalletUpdate >= 2) {
+        if (BerkeleyBatch::PeriodicFlush(dbh)) {
```

Comment:
I'm afraid this is still a layer violation, but the other way around :smile: 
- "Compacting" is berkeleyDB specific, other databases may not have anything similar
- I don't think direct use of `BerkeleyBatch` is supposed to leak into `wallet.cpp`. At least, this is the first.

Reply:
Yeah that makes sense. Thanks